Koans/AboutReflection.cs

```diff
@@ -0,0 +1,122 @@
+	// You can use reflection to create type instances at run time, and to invoke and access them
+	// Assemblies contain modules, modules contin types and types contain members.
+	// Reflection provides objects that encapsulate assemblies, modules and types.
+	// You can use reflection to dynamically create an instance of a type, bind the type to an existing object,
+	// or get the type from an existing object.
+	// You can then invoke the type's methods or access its fields and properties.
+
+	[Step(1)]
+	public void GettingAType()
+	{
+		// Everything in .net is derived from object
+		// object contains the method GetType(), which will return its type
+		// this is the most basic form of reflection
+
+		object thisIsAStringType = "a string";
+		Assert.Equal(typeof(FillMeIn), thisIsAStringType.GetType());
+	}
+
+	[Step(2)]
+	public void ViewingTheAssemblyOfAnObject()
+	{
+		// All .net projects (.csproj), when compiled are called assemblies
+		// The GetName() method will return the name of the assembly that contains this class, which will be the same as the .csproj file
+		
+		Assembly a = typeof(AboutReflection).Module.Assembly;
+
+		Assert.Equal(FILL_ME_IN, a.GetName().Name);
+	}
+
+	[Step(3)]
+	public void InspectingTheTypesInAnExternalAssembly()
+	{
+		// By now you will have noticed that DotNetKoans is cleverly using Xunit assertions as its teaching mechanic
+		// Xunit is an example of an external assembly installed via package manager
+		// Using reflection we can inspect its types
+		// Hint: we're using a type that we already know about to get the Xunit assembly and all of its types
+
+		Assembly a = typeof(Assert).Module.Assembly;
+
+		Type[] types = a.GetTypes();
+
+		Assert.Contains(typeof(FillMeIn), types);
+	}
+
+
+	[Step(4)]
+	public void ExposingNonPublicConstructor()
+	{
+		// Reflection can be dangerous!
+		// It is possible to access non-public members, types and methods in an assembly
+		// Never execute these methods or set private members as it will likely cause undefined behaviour. 
+		// They are inaccessible for a reason!
+
+		// When inspecting the Assert class (via F12/Go to Definition), we can see that there is a protected constructor without a body
+		// We can use reflection to retrieve non public constructors and execute them
+		// We specify BindingFlags to only retrieve non public, instance constructors.
+		ConstructorInfo[] constructors = typeof(Assert).GetConstructors(BindingFlags.NonPublic | BindingFlags.Instance);
+				
+		// We don't know for sure how many constructors there are in the Assert class when using reflection so we can be safe and check for null before invoking the first element in the array
+		ConstructorInfo constructor = constructors[0];
+
+		// the Invoke method returns object, but because we have retrieved this from reflecting the Assert type, we can infer that its return type is Assert
+        object assertInstance = constructor?.Invoke(null);
```

Comment:
R# tells me that the conditional access via `?` is not necessary.
This is probably the case because `System.Type` uses nullable reference types.